### PR TITLE
Initial contribution of the  bespoke scheduler infrastructure for ESH

### PR DESCRIPTION
**Pull request by kgoderis**
This is the initial contribution (and preview) of a bespoke scheduler infrastructure for ESH. It supports cron, RFC5545 and ISO8601 style expressions, and builds on top of the ThreadPoolMananger class.

It is a first step to move away from the Quartz library (not included in this PR). The contribution will allow to write code like:

    ExpressionThreadPoolExecutor scheduler = ThreadPoolManager.getExpressionScheduledPool("somePool");

Then define expressions as follows

    CronExpression exp = new CronExpression("0 15 10 ? * 6L 2015-2020");

or

    DateExpression exp = new DateExpression("2016-01-13T14:33:37+00:00");

or

    RecurrenceExpression exp = new RecurrenceExpression("FREQ=WEEKLY;INTERVAL=2;COUNT=4;BYDAY=TU,SU;WKST=MO");

and finally schedule tasks

    Runnable scheduleTask = new Runnable() {
                @Override
                public void run() {
                    logger.debug("Hello World");
                }
            };
    scheduler.schedule(scheduleTask, exp);

Signed-off-by: Karel Goderis <karel.goderis@me.com>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
+package org.eclipse.smarthome.core.common;
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java]: Add copyright info to the class

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+import java.util.Date;
+import java.util.TimeZone;
+
+public interface Expression {
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java]: Missing JavaDoc

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+     * Returns the occurrence of the recurrence rule just after a date or null
+     * if there is no occurrence date found.
+     *
+     * @param afterTime
+     *            the reference date that is just before the occurrence to
+     *            return.
+     * @return the date of the occurrence just after the date in the recurrence
+     *         set .
+     */
+    Date getTimeAfter(Date afterTime);
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java]: JavaDoc is listed twice.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    void setTimeZone(TimeZone timeZone);
+
+    /**
+     * Returns the time zone for which this <code>Expression</code>
+     * will be resolved.
+     */
+    TimeZone getTimeZone();
+
+    String getExpression();
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java]: Missing JavaDoc

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Checks if the start date is synchronized with the rule. The
+     * method throws an IllegalArgumentException if the start date is not
+     * synchronized with the rule.
+     *
+     * @param startTime
+     *            The start date to check
+     */
+    void validateStartDate(final Date startTime);
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java]: Add '@throws' clause to JavaDoc about 'IllegalArgumentException'.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java [hunk 6]
```diff
@@ -217,6 +260,107 @@ protected void afterExecute(Runnable runnable, Throwable throwable) {
...
+                Throwable cause = throwable.getCause();
+                if (cause instanceof InterruptedException) {
+                    // Ignore this, might happen when we shutdownNow() the executor. We can't
+                    // log at this point as the logging system might be stopped already.
+                    return;
+                }
+            }
+        }
+
+        public void schedule(final Runnable task, final Expression expression) {
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java]: Public method: add JavaDoc

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java [hunk 7]
```diff
@@ -217,6 +260,107 @@ protected void afterExecute(Runnable runnable, Throwable throwable) {
...
+            if (task == null) {
+                throw new NullPointerException();
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java]: I would prefer an IllegalArgumentException

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java (hunk 6 above)
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java]: Public method: check 'expression' parameter for null.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java [hunk 8]
```diff
@@ -175,6 +186,38 @@ static public ExecutorService getPool(String poolName) {
...
 
+    /**
+     * Returns an instance of an expression-driven scheduled thread pool service. If it is the first request for the
+     * given pool name, the
+     * instance is newly created.
+     *
+     * @param poolName a short name used to identify the pool, e.g. "discovery"
+     * @return an instance to use
+     */
+    static public ExpressionThreadPoolExecutor getExpressionScheduledPool(String poolName) {
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java]: 'getPool()' returns an 'ExecutorService', 'getScheduledPool()' returns an 'ScheduledExecutorService'...

I would expect that 'getExpressionScheduledPool()' returns some 'ExpressionScheduledExecutorService'. Maybe we can introduce some new interface here?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Sets the start date of the recurrence rule.
+     *
+     * @param startTime
+     *            the startDate to set
+     * @param check
+     *            True if the start date will be verified against the
+     *            rule.
+     */
+    void setStartDate(final Date startTime, final boolean check);
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java]: Not clear to me why these date methods (setStartDate(), validateStartDate() and getStartDate()) must be part of the public API.

Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java]: What does it mean "will be verified against the rule"? Which rule? By default there no such a validation?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+         *
+         * @param absMax
+         *            absolute max value.
+         * @param absMin
+         *            absolute min value.
+         * @param negativeValuesAllowed
+         *            True if negative values are allowed.
+         */
+        BoundedIntegerList(final int absMin, final int absMax, final boolean negativeValuesAllowed,
+                final boolean is1indexed) {
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java]: Missing JavaDoc for parameter 'is1Indexed'

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+ * TZD = time zone designator (Z or +hh:mm or -hh:mm)
+ *
+ * @author Karel Goderis - Intial Contribution
+ *
+ */
+public class DateExpression extends AbstractExpression {
+
+    private final static Logger logger = LoggerFactory.getLogger(DateExpression.class);
+
+    public DateExpression(final String date) throws ParseException {
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java]: JavaDoc

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        this(date, Calendar.getInstance().getTime(), TimeZone.getDefault());
+    }
+
+    private DateExpression(final String dateExpression, final Date startTime, final TimeZone zone)
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java]: Parameter 'startTime' not used.

Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java]: 'zone' can be null?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+            throws ParseException {
+        if (dateExpression == null) {
+            throw new IllegalArgumentException("date cannot be null");
+        }
+        buildExpression(dateExpression);
+        setTimeZone(zone);
+    }
+
+    public DateExpression(Expression expression) {
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java]: JavaDoc

Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java]: 'expression' can be null?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+    @Override
+    public boolean isSatisfiedBy(Date date) {
+        if (date == null) {
+            throw new IllegalArgumentException("Date cannot be null");
+        }
+        return date.equals(startDate);
+    }
+
+    public static boolean isValidExpression(String dateExpression) {
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java]: Public method: missing JavaDoc

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+        try {
+            new DateExpression(dateExpression);
+        } catch (ParseException pe) {
+            return false;
+        }
+        return true;
+    }
+
+    public static void validateExpression(String dateExpression) throws ParseException {
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java]: Really required? Use can also use isValidExpression()

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        new DateExpression(dateExpression);
+    }
+
+    @Override
+    public void validateStartDate(Date startTime) {
+        if (startTime == null) {
+            throw new IllegalArgumentException("Start time cannot be null");
+        }
```
Comment (sbussweiler) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java]: What is the goal of this method?

**Comment by kaikreuzer** [comment]
PR will be reworked, closing it for now.

**Comment by kaikreuzer** [comment]
Follow-up PR is https://github.com/eclipse/smarthome/pull/1239

**Pull request closed without merging**